locatr.go

```diff
@@ -298,40 +211,151 @@ func (l *BaseLocatr) getValidLocator(locators []string) (string, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to get response from LLM: %v", err)
 	}
 
 	llmLocatorOutput := &llmLocatorOutputDto{
 		completionResponse: *llmResponse,
 	}
+
+	l.logger.Debug(fmt.Sprintf("LLM response: %s", llmResponse.Completion))
+
+	llmResponse.Completion = fixLLmJson(llmResponse.Completion)
+
+	l.logger.Debug(fmt.Sprintf("Repaired LLM response: %s", llmResponse.Completion))
+
 	if err = json.Unmarshal([]byte(llmResponse.Completion), llmLocatorOutput); err != nil {
 		return nil, fmt.Errorf("failed to unmarshal llmLocatorOutputDto json: %v", err)
 	}
-
 	return llmLocatorOutput, nil
 }
+func (l *BaseLocatr) getLocatrOutput(htmlDOM string, userReq string) (*locatrOutputDto, error) {
+	result, err := l.llmGetElementId(htmlDOM, userReq)
+	if err != nil {
+		return nil, err
+	}
+	endAt := time.Now()
+	if result.Error == "" {
+		return &locatrOutputDto{
+			llmLocatorOutputDto:      *result,
+			LocatrRequestCompletedAt: endAt.Format(time.RFC3339),
+		}, nil
+	}
+	return nil, ErrLocatrRetrievalFailed
+}
+func (l *BaseLocatr) locateElementId(htmlDOM string, userReq string) ([]locatrOutputDto, error) {
+	llmOutputs := []locatrOutputDto{}
+	requestInitiatedAt := time.Now()
+	if l.reRankClient == nil {
+		l.logger.Debug("No rerank client setup sending full dom to llm.")
+		result, err := l.getLocatrOutput(htmlDOM, userReq)
+		if err != nil {
+			return llmOutputs, err
+		}
+		result.LocatrRequestInitiatedAt = requestInitiatedAt.Format(time.RFC3339)
+		llmOutputs = append(llmOutputs, *result)
+		if result.Error == "" {
+			return llmOutputs, nil
+		}
+		return llmOutputs, ErrLocatrRetrievalFailed
+	}
+	chunks, err := l.getReRankedChunks(htmlDOM, userReq)
+	if err != nil {
+		return llmOutputs, err
+	}
+	if len(chunks) == 0 {
+		l.logger.Debug("No chunks to process")
+		return llmOutputs, ErrNoChunksToProcess
+	}
+	if len(chunks) == 1 {
+		l.logger.Debug("Only one chunk to process, sending to llm.")
+		result, err := l.getLocatrOutput(htmlDOM, userReq)
+		if err != nil {
+			return llmOutputs, err
+		}
+		result.LocatrRequestInitiatedAt = requestInitiatedAt.Format(time.RFC3339)
+		llmOutputs = append(llmOutputs, *result)
+		if result.Error == "" {
+			return llmOutputs, nil
+		}
+		return llmOutputs, ErrLocatrRetrievalFailed
+	}
+
+	var domToProcess string
+
+	for attempt := 0; attempt < 3; attempt++ {
+		switch attempt {
+		case 0:
+			domToProcess = strings.Join(chunks[0:4], "\n")
+		case 1:
+			domToProcess = strings.Join(chunks[4:], "\n")
+		default:
+			domToProcess = htmlDOM
+		}
+
+		l.logger.Debug(fmt.Sprintf("%d attempt to find locatr with reranking", attempt+1))
+		requestCompletedAt := time.Now()
+
+		result, err := l.llmGetElementId(domToProcess, userReq)
+		if err != nil {
+			return llmOutputs, err
+		}
+
+		llmOutputs = append(llmOutputs, locatrOutputDto{
+			llmLocatorOutputDto:      *result,
+			AttemptNo:                attempt,
+			LocatrRequestInitiatedAt: requestInitiatedAt.Format(time.RFC3339),
```

Comment:
Instead of convereting time object to string here, change it to string when converting the object to json. For passing the `output` object back to user; `time` should be an object, not a string.